npm/pkg/dataplane/policies/policymanager.go

```diff
@@ -94,23 +110,32 @@ func (pMgr *PolicyManager) GetPolicy(policyKey string) (*NPMNetworkPolicy, bool)
 }
 
 func (pMgr *PolicyManager) AddPolicy(policy *NPMNetworkPolicy, endpointList map[string]string) error {
-	prometheusTimer := metrics.StartNewTimer()
 	if len(policy.ACLs) == 0 {
 		klog.Infof("[DataPlane] No ACLs in policy %s to apply", policy.PolicyKey)
 		return nil
 	}
-	defer metrics.RecordACLRuleExecTime(prometheusTimer) // record execution time regardless of failure
+
+	// TODO move this validation and normalization to controller
 	normalizePolicy(policy)
 	if err := validatePolicy(policy); err != nil {
 		return npmerrors.Errorf(npmerrors.AddPolicy, false, fmt.Sprintf("couldn't add malformed policy: %s", err.Error()))
 	}
 
 	// Call actual dataplane function to apply changes
+	timer := metrics.StartNewTimer()
 	err := pMgr.addPolicy(policy, endpointList)
+	metrics.RecordACLRuleExecTime(timer) // record execution time regardless of failure
 	if err != nil {
+		// NOTE: in Linux, Prometheus metrics may be off at this point since some ACL rules may have been applied successfully
 		return npmerrors.Errorf(npmerrors.AddPolicy, false, fmt.Sprintf("failed to add policy: %v", err))
 	}
 
+	// update Prometheus metrics on success
+	// TODO optimize metrics package to allow Inc/Dec by an amount
+	for i := 0; i < policy.numACLRulesProducedInKernel(); i++ {
```

Comment:
instead of having these loops here, can we have a func metrics.IncNumAclsByNum() which takes in an int and loops inside ?